lib/nylas/v2/model.rb

```diff
@@ -0,0 +1,174 @@
+      end
+
+      def serialize(value)
+        Date.strftime("%Y-%m-%d")
+      end
+    end
+
+    Types.registry[:date] = DateType.new
+
+    class StringType < ValueType
+      # @param [Object] uses `to_s` to cast the passed in object to a string.
+      def cast(value)
+        value.to_s
+      end
+    end
+
+    Types.registry[:string] = StringType.new
+
+    class ListAttributeDefinition
+      attr_accessor :type, :exclude_when
+
+      def initialize(type:, exclude_when:)
+        self.type = type
+        self.exclude_when = exclude_when
+      end
+
+      def cast(list)
+        return [] if list.nil? || list.empty?
+        list.map { |item| Types.registry[type].cast(item) }
+      end
+    end
+
+    class AttributeDefinition
+      attr_accessor :type, :exclude_when
+      def initialize(type:, exclude_when:)
+        self.type = type
+        self.exclude_when = exclude_when
+      end
+
+      def cast(value)
+        Types.registry[type].cast(value)
+      end
+    end
+
+    class Attributes
+      attr_accessor :data, :attribute_definitions
+
+      def initialize(attribute_definitions)
+        @attribute_definitions = attribute_definitions
+        @data = Registry.new(default_attributes)
+      end
+
+      def [](key)
+        data[key]
+      end
+
+      def []=(key, value)
+        data[key] = attribute_definitions[key].cast(value)
+      end
+
+      def serialize_for(use_case: nil)
+        data.reduce({}) do |serialized_data, (key, value)|
+          serialized_data[key] = attribute_definitions[key].cast(value)
+          serialized_data
+        end
+      end
+
+      private def default_attributes
+        attribute_definitions.keys.zip([]).to_h
+      end
+    end
+
+    module Model
+      def self.included(model)
+        model.extend(ClassMethods)
+      end
+
+      def initialize(**initial_data)
+        initial_data.each do |attribute_name, value|
+          self.send(:"#{attribute_name}=", value)
+        end
+      end
+
+      def attributes
+        @attributes ||= Attributes.new(self.class.attribute_definitions)
+      end
+
+      # @return [Hash] JSON representation of the contact. See {http://example.com/ API documentation}
+      def as_json(use_case: nil)
+        attributes.serialize_for(use_case: use_case).to_h
+      end
+
+      module ClassMethods
+        def has_n_of_attribute(name, type, exclude_when: [])
+          attribute_definitions[name] = ListAttributeDefinition.new(type: type, exclude_when: exclude_when)
+          define_accessors(name)
+        end
+
+        def attribute(name, type, exclude_when: [])
+          attribute_definitions[name] = AttributeDefinition.new(type: type, exclude_when: exclude_when)
+          define_accessors(name)
+        end
+
+        def define_accessors(name)
+          define_method :"#{name}" do
+            attributes[name]
+          end
+
+          define_method :"#{name}=" do |value|
+            attributes[name] = value
+          end
```

Comment:
We'll want to make sure we make YARD recognize that these methods are defined when attribute and has_n_of_attribute are called.